File: src/main/java/algorithm/SortBy.java
package algorithm;

public enum SortBy {

	Closer {
        @Override
        Pair choose(Pair pairA, Pair pairB) {
            return pairA.closestInAge(pairB);
        }
    },

    Further {
        @Override
        Pair choose(Pair pairA, Pair pairB) {
            return pairA.furthestInAge(pairB);
        }
    },

    LongestCombinedName {
        @Override
        Pair choose(Pair pairA, Pair pairB) {
            return pairA.longestCombinedName(pairB);
        }
    };

    abstract Pair choose(Pair pairA, Pair pairB);
}

File: src/main/java/algorithm/SortBy.kt
package algorithm

enum class SortBy {

    Closer {
        override fun choose(pairA: Pair, pairB: Pair): Pair {
            return pairA.closestInAge(pairB)
        }
    },

    Further {
        override fun choose(pairA: Pair, pairB: Pair): Pair {
            return pairA.furthestInAge(pairB)
        }
    },

    LongestCombinedName {
        override fun choose(pairA: Pair, pairB: Pair): Pair {
            return pairA.longestCombinedName(pairB)
        }
    };

    internal abstract fun choose(pairA: Pair, pairB: Pair): Pair
}
